Self-serve billing enrollment & management

paolodamico opened the pull request:
Note: https://github.com/PostHog/posthog-production/pull/36 **absolutely required before merging.** Tests will not run on CI due to this, only run tests locally until the referenced PR is merged to `master`.

## Changes

This PR introduces the following changes:

- Will show a custom message on the billing toolbar CTA [if custom message is set on the backend], otherwise it'll default to a generic message (see screenshot below).
![image](https://user-images.githubusercontent.com/5864173/94426106-facd4700-0184-11eb-9035-218dfcfd3579.png)

**Introducing the beta billing page with the following features:**
- Shows the organization's current event usage (with a color-coded indicator on usage for: normal, warning, danger: unlimited) [screenshots below].

![image](https://user-images.githubusercontent.com/5864173/94426375-657e8280-0185-11eb-8c8b-2a7c445554de.png)

![image](https://user-images.githubusercontent.com/5864173/94426421-73cc9e80-0185-11eb-8469-fdc913b08ce8.png)

![image](https://user-images.githubusercontent.com/5864173/94426431-77602580-0185-11eb-87c1-0d9057fd16b4.png)

- Allows a user whose organization is not yet in an active subscription plan to sign up for a self-serve plan. _Also includes a placeholder image for plans that have no image._

![image](https://user-images.githubusercontent.com/5864173/94426565-ad9da500-0185-11eb-8eac-587a54839f37.png)

- Allows a user to continue the billing subscription process if it hasn't been completed, or subscribe to a completely different plan.

![image](https://user-images.githubusercontent.com/5864173/94426759-fce3d580-0185-11eb-9481-12573829bc66.png)

- Shows the current billing plan for the organization with instructions to reach out in case of wanting to change or unsubscribe from the current plan.

![image](https://user-images.githubusercontent.com/5864173/94426852-1d139480-0186-11eb-816f-072c0cd6afbe.png)

- Includes a fallback in case usage information cannot be obtained from the backend.

![image](https://user-images.githubusercontent.com/5864173/94426974-47fde880-0186-11eb-89ad-0c449e709312.png)


## Checklist

- [X] All querysets/queries filter by Team (if this PR affects any querysets/queries)
- [X] Backend tests (if this PR affects the backend)
- [ ] Cypress E2E tests (if this PR affects the front and/or backend)

paolodamico committed:
billing toolbar custom message

paolodamico committed:
basic billing page

paolodamico committed:
current usage full implementation

paolodamico committed:
show billing plans

paolodamico committed:
billing plan signup

paolodamico committed:
continue setting up billing information

paolodamico committed:
convert to typescript

paolodamico commented:
@mariusandra your input particularly on the frontend would be appreciated!

paolodamico committed:
Merge branch 'master' into 200911-billing-adjustments

mariusandra commented on frontend/src/scenes/billing/Billing.js:
```diff
@@ -0,0 +1,174 @@
+        setState({ ...state, percentage })
+    }, [user])
+
+    useEffect(() => {
+        if (billingSubscription?.subscription_url) window.location.href = billingSubscription.subscription_url
+    }, [billingSubscription])
```
This can go inside some `listener` in `billingLogic`. 

Also, is the `subscription_url` an external link? If so, using `window.location.href` makes sense here. Otherwise if it's an internal link that should be redirected to without a full browser redirect, you can use this:

```js
import { router } from 'kea-router'

router.actions.push(billingSubscription.subscription_url)
```

paolodamico replied:
It is an internal route, but a Django route, not on the React frontend, so keeping the `window.location.href`

mariusandra commented on frontend/src/scenes/billing/Billing.js:
```diff
@@ -0,0 +1,174 @@
+        subscribe(plan.key)
+    }
+
+    useEffect(() => {
+        if (!user.billing?.plan || user.billing?.should_setup_billing) loadPlans()
+        if (!user.billing?.current_usage || !user.billing.plan) return
+        if (!user.billing.plan.allowance) {
+            /* Plan is unlimited */
+            setState({ ...state, percentage: null })
+            return
+        }
+        const percentage =
+            Math.round((user.billing.current_usage.value / user.billing.plan.allowance.value) * 100) / 100
+        setState({ ...state, percentage })
+    }, [user])
```
This would be great to place inside a logic as well, in a selector that just calculates the value based on the user.

mariusandra commented on frontend/src/scenes/billing/Billing.js:
```diff
@@ -0,0 +1,174 @@
+    )
+}
+
+Billing.propTypes = {
+    user: PropTypes.object.isRequired,
+}
+
+Plan.propTypes = {
+    plan: PropTypes.object.isRequired,
+    onUpgrade: PropTypes.func.isRequired,
+}
+
+export default Billing
```
Same comment here regarding the default export and the proptypes as above with BillingToolbar

mariusandra commented on frontend/src/scenes/billing/billingLogic.ts:
```diff
@@ -0,0 +1,23 @@
+import { kea } from 'kea'
+import api from 'lib/api'
+
+export const billingLogic = kea({
```
you should import `billingLogicType` here and pass it to `kea<billingLogicType>()`

mariusandra reviewed (changes requested):
Some frontend feedback submitted :)

paolodamico committed:
extract UI logic to kea

paolodamico committed:
convert billing toolbar to typescript

Twixes reviewed (approved):
Marius reviewed frontend fairly, backend changes are limited, so 👍.

paolodamico committed:
convert billing to typescript

paolodamico committed:
Merge branch 'master' into 200911-billing-adjustments

paolodamico committed:
put billing page behind a feature flag

paolodamico commented:
Thanks for all the feedback @mariusandra super helpful!! I've addressed all your comments. I've put the page (showing it on the sidebar) behind a feature flag to be able to release without having to merge https://github.com/PostHog/posthog-production/pull/36 which is waiting on CH. Can you just review this last change pls?

mariusandra commented on frontend/src/lib/components/BillingToolbar.tsx:
```diff
@@ -0,0 +1,31 @@
+import { WarningOutlined, ToolFilled } from '@ant-design/icons'
+import { Button } from 'antd'
+
+export function BillingToolbar({
+    billingUrl = null,
+    message,
+}: {
+    billingUrl: string | null
+    message: string
+}): JSX.Element {
+    return (
+        <>
+            {billingUrl && (
+                <div className="card">
+                    <div className="card-body" style={{ display: 'flex' }}>
```
I would just add a hook here to `const { user } = useValues(userLogic)`, so there's no need for passing props

mariusandra commented on frontend/src/scenes/billing/Billing.tsx:
```diff
@@ -0,0 +1,135 @@
+        subscribe(plan.key)
+    }
+
+    useEffect(() => {
+        if (!user?.billing?.plan || user?.billing?.should_setup_billing) loadPlans()
+    }, [user])
```
This could go in the billingLogic in an `afterMount` event. All other data fetching for the billing info is there, so it just makes sense for this to be together

mariusandra commented on frontend/src/scenes/billing/Billing.tsx:
```diff
@@ -0,0 +1,135 @@
+        if (!user?.billing?.plan || user?.billing?.should_setup_billing) loadPlans()
+    }, [user])
+
+    useEffect(() => {
+        if (billingSubscription?.subscription_url) {
+            window.location.href = billingSubscription.subscription_url
+        }
+    }, [billingSubscription])
```
This could also be in the billingLogic. I'd put it as a listener for the `subscribeSuccess` event

mariusandra reviewed (changes requested):
Some changes, but overall much much better than before! 👍

paolodamico committed:
features flag logic in kea

paolodamico committed:
code improvements

paolodamico commented:
@mariusandra lmk if you recommend any additional changes, I'll merge otherwise

mariusandra commented on frontend/src/layout/Sidebar.js:
```diff
@@ -240,6 +244,15 @@ export function Sidebar({ user, sidebarCollapsed, setSidebarCollapsed }) {
                             <Link to={'/annotations'} onClick={collapseSidebar} />
                         </Menu.Item>
+
+                        {featureFlags && featureFlags['billing-management-page'] && (
```
This is all great! Nicely done!
The only nit I can give is that `featureFlags` is always defined, so no need for the `&&`. Definitely not worth opening another PR for that

paolodamico replied:
Oh you're right, thanks for pointing that out, I'll remember to fix it in my next frontend PR.
